### PR TITLE
chore(exex): emit warn log when WAL grows beyond a certain number of blocks

### LINKED ISSUE #12626: Emit a WARN log if the ExEx WAL grows beyond a certain number of blocks
### Describe the feature

If an ExEx doesn't emit `FinishedHeight` events, the WAL will never be cleared and will grow indefinitely https://github.com/paradigmxyz/reth/blob/4b4f9cf40626c8714f0a0caa152a970fa6e5dfa0/crates/exex/exex/src/manager.rs#L348-L367

To help prevent this user error, we can emit a WARN log in `ExExManager::finalize_wal` if the WAL contains more than 128 blocks after a call to `self.wal.finalize()`.

To do that, we need to add a method to `Wal` that will return the number of blocks in the WAL.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by htiennv**
Closes #12626.

### CODE REVIEW

File: crates/exex/exex/src/manager.rs [hunk 1]
```diff
@@ -377,6 +383,14 @@ where
                 .unwrap();
 
             self.wal.finalize(lowest_finished_height)?;
+            if self.wal.num_blocks() > (DEFAULT_MAX_WAL_BLOCKS_WARNING as u64) {
+                warn!(
+                    target: "exex::manager",
+                    "WAL contains {} blocks, exceeding the limit of {}",
+                    self.wal.num_blocks(),
+                    DEFAULT_MAX_WAL_BLOCKS_WARNING,
```
Comment (shekhirin) [inline crates/exex/exex/src/manager.rs:391]: I would rephrase this to something that asks the user to check that their ExEx is behaving correctly and emits `FinishedHeight` event, because technically there's no limit for the WAL size, and it's just a pure "FYI" log.

File: crates/exex/exex/src/manager.rs [hunk 2]
```diff
@@ -35,6 +35,12 @@ use tokio_util::sync::{PollSendError, PollSender, ReusableBoxFuture};
 /// or 17 minutes of 1-second blocks.
 pub const DEFAULT_EXEX_MANAGER_CAPACITY: usize = 1024;
 
+/// The default maximum number of blocks allowed in the WAL before emitting a warning.
+///
+/// This constant defines the threshold for the Write-Ahead Log (WAL) size. If the number of blocks
+/// in the WAL exceeds this limit, a warning is logged to indicate potential issues.
+pub const DEFAULT_MAX_WAL_BLOCKS_WARNING: usize = 128;
```
Comment (shekhirin) [inline crates/exex/exex/src/manager.rs:42]: I don't think we want to make it configurable, so let's omit the "default" part
```suggestion
/// The maximum number of blocks allowed in the WAL before emitting a warning.
///
/// This constant defines the threshold for the Write-Ahead Log (WAL) size. If the number of blocks
/// in the WAL exceeds this limit, a warning is logged to indicate potential issues.
pub const WAL_BLOCKS_WARNING: usize = 128;
```

**Review by shekhirin — CHANGES_REQUESTED** [review]
Amazing, that should work! Only a couple nits

File: crates/exex/exex/src/manager.rs (hunk 1 above)
Comment (htiennv) [reply crates/exex/exex/src/manager.rs]: Thanks @shekhirin for your reviewing. Can you suggest me a warning message. I will fix it in the next commit.

File: crates/exex/exex/src/manager.rs [hunk 3]
```diff
@@ -377,6 +383,14 @@ where
                 .unwrap();
 
             self.wal.finalize(lowest_finished_height)?;
+            if self.wal.num_blocks() > (WAL_BLOCKS_WARNING as u64) {
+                warn!(
+                    target: "exex::manager",
+                    "WAL contains {} blocks, exceeding the limit of {}",
+                    self.wal.num_blocks(),
+                    WAL_BLOCKS_WARNING,
+                );
```
Comment (shekhirin) [inline crates/exex/exex/src/manager.rs:392]:
```suggestion
                warn!(
                    target: "exex::manager",
                    blocks = ?self.wal.num_blocks(),
                    "WAL contains too many blocks and is not getting cleared. That will lead to increased disk space usage. Check that you emit the FinishedHeight event from your ExExes."
                );
```

Comment (htiennv) [reply crates/exex/exex/src/manager.rs:392]: Thanks for your help!

File: crates/exex/exex/src/wal/cache.rs [hunk 4]
```diff
@@ -35,6 +35,11 @@ impl BlockCache {
         self.notification_max_blocks.is_empty()
     }
 
+    /// Returns the number of blocks in the cache.
+    pub(super) fn num_blocks(&self) -> u64 {
+        self.committed_blocks.len() as u64
```
Comment (shekhirin) [inline crates/exex/exex/src/wal/cache.rs:40]: nit: we can remove a bunch of conversions by returning `usize` here

Comment (htiennv) [reply crates/exex/exex/src/wal/cache.rs:40]: yeah, i agree

**Review by shekhirin — APPROVED** [review]
Thank you!
